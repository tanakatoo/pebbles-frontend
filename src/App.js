import './styles/app.css';
import NavBar from "./components/common/NavBar"
import "./styles/app.css"
import { useSelector } from 'react-redux';
import MyRoutes from "./MyRoutes"

function App() {
  //get the font from the store
  const font = useSelector(state => state.langFont.font)

  //we have to hardcode the string of the font because tailwind doesn't know what to use at build
  return (
    <div className={`App ${font == 'font-poppins' ? 'font-poppins' : 'font-notoJP'}`}>
      <NavBar />
      <MyRoutes />
    </div >
  );
}

export default App;
