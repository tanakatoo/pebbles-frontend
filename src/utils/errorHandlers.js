import { rest } from 'msw'

let loginResponse = {
    token: 'abcdefg',
    user: {
        id: 2,
        username: 'hello',
        role: 'regular',
        premium_join_date: '',
        premium_end_date: '',
        free_trial_start_date: '',
    }
}



export const errorHandlers = [

    //login
    rest.post('http://localhost:3001/auth/login', (req, res, ctx) => {
        const failLogin = {
            "error": {
                "message": "INVALID_CREDENTIALS",
                "status": 401
            }
        }
        console.log(res(ctx.json([failLogin])))
        return res(ctx.json([failLogin]));
    }),

    rest.post('http://localhost:3001/email', (req, res, ctx) => {
        console.log('MOCK error /email');
        const noData = {
            "error": {
                "message": "no data",
                "status": 400
            }
        }
        return res(ctx.json(noData));
    }),

]