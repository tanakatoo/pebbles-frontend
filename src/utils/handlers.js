import { rest } from 'msw'

let loginResponse = {
    token: 'abcdefg',
    user: {
        id: 2,
        username: 'hello',
        role: 'regular',
        premium_join_date: '',
        premium_end_date: '',
        free_trial_start_date: '',

    }
}


export const handlers = [
    rest.post('http://localhost:3001/auth/login', (req, res, ctx) => {


        console.log(res(ctx.json(loginResponse)))
        return res(
            // Respond with a 200 status code
            // ctx.status(200),
            res(ctx.json(loginResponse))
        )
    }),

    // Called via <ExploreCommunity type="studyBuddy" />
    rest.get('http://localhost:3001/study-buddies/1', (req, res, ctx) => {
        console.log('MOCK /study-buddies/1');
        console.log('  response:', {});
        return res(ctx.json({}));
    })
]