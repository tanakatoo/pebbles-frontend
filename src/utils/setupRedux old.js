
import { configureStore } from '@reduxjs/toolkit';
import rootReducer from "../reducers/rootReducer"
import { persistStore, persistReducer, FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER } from 'redux-persist';

import storage from 'redux-persist/lib/storage'
import autoMergeLevel2 from 'redux-persist/lib/stateReconciler/autoMergeLevel2'

export const persistConfig = {
    key: 'root', // Key prefix for the persisted state
    storage, // Storage mechanism (e.g., local storage)
    stateReconciler: autoMergeLevel2, // Specify which parts of the state to persist
    whitelist: ['profile', 'langFont']
};

export const persistedReducer = persistReducer(persistConfig, rootReducer);
export const store = configureStore({
    reducer: persistedReducer,
    middleware: (getDefaultMiddleware) =>
        getDefaultMiddleware({
            serializableCheck: {
                ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER]
            }
        })
})
export const persistedStore = persistStore(store)