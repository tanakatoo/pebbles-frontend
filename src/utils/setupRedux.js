import React from "react";
import { render } from "@testing-library/react";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
// As a basic setup, import your same slice reducers
import { persistStore, persistReducer, FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER } from 'redux-persist';
import { PersistGate } from 'redux-persist/integration/react'
import { ListSlice } from "../reducers/rootReducer";
import rootReducer from "../reducers/rootReducer"
import storage from 'redux-persist/lib/storage'
import storeSetup from "./storeSetup"
import autoMergeLevel2 from 'redux-persist/lib/stateReconciler/autoMergeLevel2'

// const persistConfig = {
//     key: 'root', // Key prefix for the persisted state
//     storage, // Storage mechanism (e.g., local storage)
//     stateReconciler: autoMergeLevel2, // Specify which parts of the state to persist
//     whitelist: ['profile', 'langFont']
// };
// const persistedReducer = persistReducer(persistConfig, rootReducer);

// const persistedStore = persistStore(store)
// const root = ReactDOM.createRoot(document.getElementById('root'));


export function renderWithProviders(
    ui,
    {
        preloadedState = {},
        // Automatically create a store instance if no store was passed in
        store = storeSetup(preloadedState),
        ...renderOptions
    } = {}
) {
    function Wrapper({ children }) {

        return <Provider store={store}>{children}</Provider>;
    }

    // Return an object with the store and all of RTL's query functions
    return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}