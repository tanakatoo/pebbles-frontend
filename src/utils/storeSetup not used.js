import { combineReducers, configureStore } from '@reduxjs/toolkit'

import rootReducer from '../reducers/rootReducer'
// import userReducer from '../features/users/userSlice'

// Create the root reducer separately so we can extract the RootState type
// const rootReducer = combineReducers({
//     user: userReducer
// })

export const setupStore = preloadedState => {
    return configureStore({
        reducer: rootReducer,
        preloadedState
    })
}