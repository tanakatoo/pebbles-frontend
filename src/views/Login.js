import React, { useState, useContext } from "react"
import usePageText from "../hooks/usePageText";
import { useDispatch, useSelector } from "react-redux";
import ServerError from "../components/form/ServerError";
import loginSchema from "../components/form/validation/loginSchema";
import { actionLogin } from "../reducers/actionCreator";
// import { TextInput } from "../components/form/Fields";
import Input from "../components/form/Input";
import { Formik, Form } from "formik"
import { Button } from "../components/button/Button";
import AuthApi from "../api/auth";
import { useNavigate } from "react-router-dom";
import PebblesApi from "../api/base";
import { actionSetMsg } from "../reducers/actionCreator";
// import FlashMessageContext from "../contexts/FlashMessageContext";
import CTA from "../components/common/CTA";
import { Link } from "react-router-dom";

const Login = () => {
    const dispatch = useDispatch()
    const [errors, setErrors] = useState([])
    const navigate = useNavigate()
    const [pageText, lang] = usePageText("login")
    // const setFlashMessage = useContext(FlashMessageContext)
    const token = useSelector(state => state.profile.token)

    const INITIAL_DATA = {
        username: '',
        password: ''
    }

    console.log('pebbles api token is', PebblesApi.token)
    if (token) {
        //should go to their dashboard
        navigate('/users/dashboard')
    }
    console.log(errors)
    return (
        <>
            <CTA msg={pageText.CTA} msgBtn={pageText.CTA_BTN} btnLink={pageText.CTA_LINK} />
            <div className="mt-8 flex flex-col justify-center items-center mx-8">
                {Object.keys(errors).length > 0 && <ServerError msg={errors} title={pageText.ERROR_TITLE} />}
                {errors.length === 0 && <h1 className="text-center mb-[56px] text-mobile-header-2">{pageText.H1}</h1>}
                <Formik
                    initialValues={INITIAL_DATA}
                    validationSchema={loginSchema}
                    onSubmit={async (values, { setSubmitting }) => {

                        setErrors([])
                        try {
                            const res = await AuthApi.login(values.username, values.password)
                            //call dispatch to set token in profileReducer
                            dispatch(actionLogin(res)) //save token and then profile
                            // setFlashMessage('LOGIN')
                            navigate('/users/dashboard')
                        } catch (e) {
                            if (e instanceof TypeError) {
                                //means server is down
                                setErrors(["UNKNOWN"])
                            } else {

                                setErrors(e)
                            }
                        } finally {
                            setSubmitting(false)
                        }


                    }}

                >{formik => {
                    return (
                        <Form className="flex flex-col max-w-[500px]">

                            <Input
                                label={pageText.USERNAME}
                                name="username"
                                type="text"
                            />

                            <Input
                                label={pageText.PASSWORD}
                                name="password"
                                type="password"
                                placeholder=""
                            />
                            <Button btnText={pageText.SUBMIT}
                                type="submit"
                                extraClasses="mt-12"
                                lang={lang}
                                isSubmitting={formik.isSubmitting} />
                            <Button bkColor="bg-white"
                                textColor="text-primary"
                                btnText={pageText.CREATE_ACCT}
                                type="button"
                                lang={lang}
                                extraClasses="mt-8 border mb-[76px]"
                                isSubmitting={formik.isSubmitting}
                                link="/register" />
                        </Form>
                    )
                }}
                </Formik>
            </div>
        </>
    );

};

export default Login;
