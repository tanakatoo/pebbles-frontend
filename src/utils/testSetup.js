// import { persistedStore, store } from './setupRedux';
import { render } from '@testing-library/react'
import { Provider } from 'react-redux';
import { PersistGate } from 'redux-persist/integration/react'
import { setupTestStore } from './setupRedux';


const renderWithProviders = (ui, options = {}) => {
    const [store, persistedStore] = setupTestStore(options.preloadedState);

    const AllProviders = ({ children }) => {
        return (
            <Provider store={store}>
                <PersistGate loading={null} persistor={persistedStore}>
                    {children}
                </PersistGate>
            </Provider>
        )
    };

    return { store, ...render(ui, { wrapper: AllProviders, ...options }) };
}


// "re-export everything"
export * from '@testing-library/react';

export { renderWithProviders };