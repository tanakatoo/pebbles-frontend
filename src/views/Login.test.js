import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { fireEvent, renderWithProviders } from '../utils/testSetup';
import Login from './Login';
import App from '../App';
import { waitFor } from '../utils/testSetup';
import { act } from 'react-dom/test-utils';


// Mock window.scrollTo()
window.scrollTo = jest.fn();
afterAll(() => {
    jest.clearAllMocks();
});

beforeEach(async () => {
    const { getByTestId, getByLabelText, getByText } = renderWithProviders(
        <MemoryRouter >
            <App />
        </MemoryRouter>
    );

    // Populate form
    let changeLang = getByText('EN', { exact: true });
    fireEvent.click(changeLang);
})

describe('Login', () => {
    test('renders without crashing', () => {
        renderWithProviders(
            <MemoryRouter>
                <Login />
            </MemoryRouter>
        );
    });

    test('matches snapshot', async () => {
        const { asFragment } = renderWithProviders(
            <MemoryRouter>
                <Login />
            </MemoryRouter>
        );

        expect(asFragment()).toMatchSnapshot();
    });

    test('valid user login', async () => {
        const { getByTestId, getByLabelText, getByText } = renderWithProviders(
            <MemoryRouter initialEntries={["/login"]}>
                <App />
            </MemoryRouter>
        );


        let elem = getByLabelText('Username or email', { exact: false });
        fireEvent.change(elem, { target: { value: 'hello' } });
        elem = getByLabelText('Password', { exact: false });
        fireEvent.change(elem, { target: { value: 'asdfasdf' } });

        // Submit form
        // act(() => {
        //     fireEvent.click(getByTestId(/loginFormLogin/));
        // })
        fireEvent.click(getByTestId(/loginFormLogin/));

        // Wait for pageText to be populated
        await waitFor(() => {
            expect(getByTestId('wait-for-pagetext')).not.toBeEmptyDOMElement();
        });

        // Are we on the Dashboard page?
        expect(getByText(/Learn how to use Pebbles/)).toBeInTheDocument();
    });
});