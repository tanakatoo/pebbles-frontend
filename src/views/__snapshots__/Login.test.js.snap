// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex w-full bg-primary text-white px-4 py-3 mb-8 justify-center items-center"
  >
    <p
      class="mr-8 font-medium"
    >
      Practice and learn Japanese
    </p>
    <a
      href="/study-buddies"
    >
      <button
        class="drop-shadow 
                inline-flex
                items-center
                justify-center
                gap-2
                bg-white 
                text-primary-dark 
                py-2 
                px-4 
                rounded-xl
                disabled:opacity-50 text-mobile-card-header font-medium}
                hover:opacity-75"
        data-testid=""
        type="button"
      >
        <span
          class="font-medium"
        >
          Study buddies
        </span>
      </button>
    </a>
  </div>
  <div
    class="mt-8 flex flex-col justify-center items-center mx-8"
  >
    <h1
      class="text-center mb-[56px] text-mobile-header-2"
    >
      Welcome back
    </h1>
    <form
      action="#"
      class="flex flex-col w-full max-w-[500px]"
    >
      <div
        class="flex flex-col w-full"
      >
        <label
          class="mb-2 
            text-mobile-section-header
            font-medium"
          for="username"
        >
          Username or email
        </label>
        <input
          class="mb-2 
            rounded-ml py-3 px-4 text-black
             placeholder-gray
             undefined"
          id="username"
          name="username"
          type="text"
          value=""
        />
      </div>
      <div
        class="flex flex-col w-full"
      >
        <label
          class="mb-2 
            text-mobile-section-header
            font-medium"
          for="password"
        >
          Password
        </label>
        <input
          class="mb-2 
            rounded-ml py-3 px-4 text-black
             placeholder-gray
             undefined"
          id="password"
          name="password"
          placeholder=""
          type="password"
          value=""
        />
      </div>
      <button
        class="drop-shadow 
                inline-flex
                items-center
                justify-center
                gap-2
                bg-primary border-primary 
                text-white 
                py-3 
                px-4 
                rounded-xl
                disabled:opacity-50 mt-12
                hover:opacity-75"
        data-testid="data-testid=loginFormLogin"
        type="submit"
      >
        <span
          class="font-medium"
        >
          Login
        </span>
      </button>
      <p
        class="mt-2 text-center text-gray-text"
      >
        Forgot password?
        <span
          class="text-link hover:text-primary"
        >
          <a
            href="/change-password"
          >
             Click here to reset
          </a>
        </span>
      </p>
      <button
        class="drop-shadow 
                inline-flex
                items-center
                justify-center
                gap-2
                bg-white 
                text-primary 
                py-3 
                px-4 
                rounded-xl
                disabled:opacity-50 mt-8 border mb-[76px]
                hover:opacity-75"
        data-testid=""
        type="button"
      >
        <a
          class="font-medium"
          href="/register"
        >
          Register free account
        </a>
      </button>
    </form>
  </div>
</DocumentFragment>
`;
